firedrake/assemble_expressions.py

```diff
@@ -311,10 +294,9 @@ def fast_key(self):
 
     @cached_property
     def slow_key(self):
-        """A slow lookup key for this expression (relabelling UFL indices)."""
-        self.relabeller._reset()
-        rvalue, = map_expr_dags(self.relabeller, [self.rvalue])
-        return (type(self), hash(self.lvalue), hash(rvalue))
+        """A slow lookup key for this expression."""
+        from firedrake.interpolation import hash_expr
+        return (type(self), hash(self.lvalue), hash_expr(self.rvalue))
```

Comment:
probably hash(lvalue) is not stable for two coefficients that would be disk cache identical. I never considered using this for an on disk cache, only in memory

Reply:
To be perfectly honest I think a disk cache here is probably overkill anyway. I could easily switch to using only an in-memory cache here. WDYT?

Reply:
Sure, I forget exactly how the cache-lookup proceeds but I vaguely recall that the necessary thing is that the spaces match (this hashes on the coefficient in the left hand side of the assign rather than the space)